src/expr.jl

```diff
@@ -2,25 +2,10 @@
 # Builtin Expr type.
 
 istree(x::Expr) = true
-exprhead(e::Expr) = e.head
 
-operation(e::Expr) = expr_operation(e, Val{exprhead(e)}())
-arguments(e::Expr) = expr_arguments(e, Val{exprhead(e)}())
+operation(e::Expr) = e.head
+arguments(e::Expr) = e.args
 
```

Comment:
Here comes my consideration on this:

Reply:
@willow-ahrens  

I agree that this feels like the most natural way of handling Expr. I actually
like this, but I would like to ask for your consideration and help on the design
of the Metatheory/SymbolicUtils pattern matcher, which is exactly the same code.
(Metatheory just extends it with e-graphs, but classical term rewriting is
exactly copy-pasted over from SU).

As we discussed yesterday in the call, most pattern matchers in languages like
OCaml and friends, actually match on the type of the node.

```ocaml
type 'a tree =
| Leaf
| Node of 'a * 'a tree * 'a tree

let rec size = function
  | Leaf -> 0
  | Node (_, l, r) -> 1 + size l + size r
```

This is actually very nice, in Julia this is implemented by the amazing library
https://github.com/thautwarm/MLStyle.jl I've used MLStyle extensively for years
now.

To match against an `Expr` you pattern match against its constructor. This is
nice if you consider symbolic manipulation with ADTs and plays well with Julia
structs.

```julia
julia> using MLStyle

julia> ex = :(foo(a,b,c))
:(foo(a, b, c))

julia> @match ex begin 
           Expr(:call, :foo, args...) => Expr(:call, :bar, reverse(args)...)
           _ => error()
       end
:(bar(c, b, a))
```

It's actually quite easy to replicate "@rule" behavior

```julia
# @rule f(~x, ~y) => g(~y, ~x)  
(x -> @match x begin 
    Expr(:call, :f, x, y) => Expr(:call, :g, y, x)
    Term(f, x, y) => Term(g, y, x) # f and g need be defined as in SU
    _ => nothing
)
```

### Point 1: Difference with SU and MT matcher

In the context of SU and MT, things have always been a little different. The issue is
that MLStyle pattern matcher rules **depend on the types of terms** - we want our
MT/SU rewrite rule patterns to not depend on the type of terms, but **to be general**, as long as 
**TermInterface is implemented**, and that's the
main motivation behind this library, thus SU/MT's `@rule lhs => rhs` should
match on lhs the same way on any term type `T` that support TermInterface.

I think we should achieve unanimous agreement to proceed in order to make things
easy for implementers of symbolic types, without losing generality - after we
agreed that `operation(e::Expr) = e.head` and same for `args`.

We can lift this problem away 1 layer, from TermInterface to MT/SU matchers,
or consider a nice **trait** that can help us consider both styles.

### The problem - Operation as node or operation as first child

AKA **Maintaining the same rule language on term types where function calls in the form f(xs...) are represented by `operation(t) = :call` or `operation(t) = f`**

Let's consider a simple rule and rewriting example. I'm using `-->` to denote
symbolic substitution, instead of `=>` rules where the RHS is evaluated. This is
only because the example can apply to both MT and SU, as in this example we care
only about how we are matching the LHS `f(a,b)`, and not really about what happens after.

At the moment, pattern matching logic is the same for MT and SU: match the
operation, if it matches, loop over matching the arguments, binding variables
when encountered.

Please note that currently, in SU - the interface for Expr is broken.
https://github.com/JuliaSymbolics/SymbolicUtils.jl/blob/master/test/interface.jl


```julia
r = @rule f(~a, ~b) --> g(b,a)

# SU example

function f end 
function g end
@syms j k 
t = f(j,k)     # SU term
operation(t) == f
arguments(t) == [j, k]

r(t) == g(k,j)

# Expr example (real world example, we may inside a CodeInfo)

ex = :(f(x,y))
ex == Expr(:call, :f, :x, :y) 
operation(ex) == :call 
arguments(ex) == [:f, :x, :y]

r(ex) # ??? nothing
```

I think it's quite easy to see that having `r(ex) == nothing` feels unnatural, and
will most likely confuse newcomers, or require people to write rewrite rules
many times.


### Point 2. Rewrite rules and our matcher should work on anything that implements TermInterface

It's the whole point of this package. If we want require users to write 
`@rule f(~a,~b) => g(~b,~c)` for Symbolics terms, then `@rule Expr(:call, :f ~a,~b) => Expr(:call, :g, ~b,~c)` 
for `Expr`, or even worse `@rule :call(:f, ~a,~b) => :call(:g, ~b,~c)`. 
It would then make no sense to try to maintain this package at all,
as MLStyle (pointed above) is already doing everything we need and SU rewrites for SU.

Having **"one rule language to rule them all"** was the cool thing!

### Point 3. We should rely on dispatch for this

We agreed in the call that dispatching on types for `maketerm/similarterm` is the way to go. 
It should be (as it always was in MT/SU) the same for pattern matching.

I hope everybody agrees that with `r = @rule f(~a, ~b) --> g(~b, ~a)`, 
calling `r(:(f(x,y)))` should return `:(g(y,x))`

### Point 4. The context of patterns.

If we consider the construction of rules in an abstract way, in both MT and SU
we have a `@rule : Expr -> (Pattern, Union{Function, Pattern})` macro that
always makes a pattern of the LHS. SU discards `Expr.head` information and
builds a `Term`, which only works when the functions are defined. MT does the same
thing but allows for maintaining the original `Expr` information (metadata, expr.head), and to quote things that are not defined (done at macro expansion in local scope
with no overhead, if `f` is defined where macro is called, then it's function object, symbol otherwise).
It's stored in a `PatTerm`.

It's my goal to have MT pattern matcher work seamlessly and **in an intuitive way** with any expression type, 
requiring users to write rules only once.

### Solutions

Let's go through the different redesign proposals and why we rejected them.

1. ~~exprhead~~ - `operation` is always the "function being called", and `exprhead` is defined to maintain the structure of the `Expr -> Rule` mapping in case `Expr` is matched. Required a kwarg in `similarterm/maketerm`. **Abstraction Leak**.
2. ~~Define `Head` types~~. We agreed that it's a burden on the user (needs to define many types) and that `operation,arguments,similarterm` can be done by dispatching on term type `T` (`Expr,` `BasicSymbolic`). **Complicates users' implementation of TI**. Doesn't solve the OP as node/first child problem.
3. ~~`Head` types with traits~~. Even more complex, was my idea to define a trait to handle where the "operation" should reside.
4. ~~`is_function_call`~~. Boils down the issue of `exprhead` to a single boolean. Small and elegant, as (1) it required a `kwarg` in `maketerm/similarterm`. **Abstraction Leak**.
5. **Where we are now**: Nice and minimal implementation. `similarterm` dispatches on type. I'm afraid that unless we propose something for handling `operation(ex::Expr) = ex.head` in pattern matching it will break the "write rule once us it anywhere" assumption that is (or should be) the strength point of MT and SU.

---

Reply:
I'm trying to put together **a solution that involves traits on type T**, and avoids extra kwargs to `similarterm`. 

I'm afraid it's gonna be complicated. Please let me know if you have any ideas. Once we unlock this I'll proceed with releasing MT 3.0 and integrating it with Symbolics.jl. Folks from rust have started https://egraphs.org/ and I've been invited to present. Submissions due Apr 5. If we come to a solution by then, you're more than welcome to coauthor.

Please let's also consider that `is_function_call(::T)::Bool` was super minimal, easy to implement for users of TermInterface and enough to make the matcher work all the times and fast. 

In order to avoid the extra kwarg `is_call` being always specified in `similarterm`, we could easily allow to define `is_function_call(::Type{T})::Bool`, such that

```julia
function similarterm(x::T, op, args; is_call=is_function_call(T), symtype = nothing, metadata = nothing) where {T}
  similarterm(T, op, args; is_call, symtype, metadata)
end

function similarterm(x::Type, op, args; is_call=is_function_call(x), symtype = nothing, metadata = nothing)
  # is_call this would always be true with BasicSymbolic, but can be set to 'false'  when op::typeof(getindex)
end
```

Reply:
> I hope everybody agrees that with r = @rule f(~a, ~b) --> g(~b, ~a), calling r(:(f(x,y))) should return :(g(y,x))

On the call, Shashi and I did not agree.

SymbolicUtils provides the `term` type, easily constructible with `@syms`, to work with a function-calls-as-operations AST.

What if we provide the following struct in TermInterface and explain how to use it?

```
struct WithCall
    expr::Expr
end

function operation(wc::WithCall)
    expr = wc.expr
    if expr.head == :call
        return expr.args[1]
    else
        return expr.head
    end
end

function arguments(wc::WithCall)
    expr = wc.expr
    if expr.head == :call
        return @view expr.args[2:end]
    else
        return expr.args
    end
end

function similarterm(::Type{WithCall}, op, args)
    if op isa Symbol
        return WithCall(Expr(op, args...))
    else
        return WithCall(Expr(:call, op, args...))
    end
end
```

Reply:
Then, we would have `r(WithCall(:(f(x,y))))` should return `WithCall(:(g(y,x)))`

We could call this struct something more intuitive, perhaps, like "ExprCall"?